controllers/extensionRequests.js

```diff
@@ -14,73 +14,166 @@ const { parseQueryParams } = require("../utils/queryParser");
+      }
 
-    if (req.userData.id !== extensionBody.assignee && !req.userData.roles?.super_user) {
-      return res.boom.forbidden("Only assigned user and super user can create an extension request for this task.");
-    }
+      if (!assigneeId) {
+        return res.boom.badRequest("User with this id or username doesn't exist.");
+      }
 
-    const { taskData: task } = await tasks.fetchTask(extensionBody.taskId);
-    if (!task) {
-      return res.boom.badRequest("Task with this id or taskid doesn't exist.");
-    }
-    if (task.assignee !== assigneeUsername) {
-      return res.boom.badRequest("This task is assigned to some different user");
-    }
-    if (task.endsOn >= extensionBody.newEndsOn) {
-      return res.boom.badRequest("The value for newEndsOn should be greater than the previous ETA");
-    }
-    if (extensionBody.oldEndsOn !== task.endsOn) {
-      extensionBody.oldEndsOn = task.endsOn;
-    }
+      if (req.userData.id !== extensionBody.assignee && !req.userData.roles?.super_user) {
+        return res.boom.forbidden("Only assigned user and super user can create an extension request for this task.");
+      }
 
-    const prevExtensionRequest = await extensionRequestsQuery.fetchExtensionRequests({
-      taskId: extensionBody.taskId,
-      assignee: extensionBody.assignee,
-    });
-    if (prevExtensionRequest.length) {
-      return res.boom.forbidden("An extension request for this task already exists.");
+      const { taskData: task } = await tasks.fetchTask(extensionBody.taskId);
+      if (!task) {
+        return res.boom.badRequest("Task with this id or taskid doesn't exist.");
+      }
+      if (task.assignee !== assigneeUsername) {
+        return res.boom.badRequest("This task is assigned to some different user");
+      }
+      if (task.endsOn >= extensionBody.newEndsOn) {
+        return res.boom.badRequest("The value for newEndsOn should be greater than the previous ETA");
+      }
+      if (extensionBody.oldEndsOn !== task.endsOn) {
+        extensionBody.oldEndsOn = task.endsOn;
+      }
+
+      const latestExtensionRequest = await extensionRequestsQuery.fetchLatestExtensionRequest({
+        taskId: extensionBody.taskId,
+      });
+
+      if (latestExtensionRequest[0] && latestExtensionRequest[0].status === "PENDING") {
+        return res.boom.badRequest("An extension request for this task already exists.");
+      }
+
+      // dont allow newETA<taskEndsOnETA
+      if (latestExtensionRequest[0]) {
+        if (latestExtensionRequest[0].status === "APPROVED" || latestExtensionRequest[0].status === "DENIED") {
+          if (extensionBody.newEndsOn < task.endsOn) {
+            return res.boom.badRequest("An extension request can't be equal to or lesser than taskEndsOn date");
+          }
+        }
+      }
+
+      // logic for extension request count
+      let requestNumber;
+      if (latestExtensionRequest[0] && latestExtensionRequest[0].userId === assigneeId) {
```

Comment:
- Suppose user A creates an ER for task 1. 
- task 1 gets assigned to user B
- user B raises an ER for task 1.
- task 1 gets re-assigned to user A.
- if user A raises ER again wouldn't the er request number get reset to 1 with this implementation?
- Shouldn't it be 2?

To solve this I think we should fetch the `latestExtensionRequest` based on assigneeId and taskId.

Reply:
No if assignee is different than previous and if it goes to same user I am making it ER 1.